Incorporate auto-mode into modable-buffer.

aartaka opened the pull request:
# Description

This moves all the functionality of the `auto-mode` into `modable-buffer`, which results in cleaner code and more reliable mode- and page- specific mode settings. What's new:
- `define-auto-rule` as a simple way to define arbitrary rules without touching the file with user rules.
- `auto-mode` renamed to `auto-rules` as a feature.
- Rule `test`s are now being evaluated, which means we can now put arbitrary lambdas inside the rules!
- Default rules for `gopher`, `gemini`, and all the internal pages that require it.
- `apply-all-matching-auto-rules-p` that allows applying all the matching rules instead of one (requested long ago by @hendursaga)!
- Minor fixes for `no-procrastinate-mode`, and others.
- I've reshuffled the file dependencies, and auto-rules.lisp (the renamed mode/auto.lisp file) now has access to buffer types and history, improving the typing and semantics of functions there!

What's removed:
- Numerous `enable-modes` doing the manual work.
- Lots of `rememberable-p`

Fixes #2004
Fixes #2066 
Fixes #962 

# Discussion

1. It feels suspicious that I had to remove so many `rememberable-p` from the interface-related modes. The only modes that are left with `(rememeberable-p nil)` are the real one-shot modes. Maybe rename it to `one-shot-p` then?
2. Some of the slots from `auto-mode` feel weird in `modable-buffer`. Maybe there's a better place for the slots of `auto-mode`? I was unable to think of any...
3. I am now suspicious of any `enable-modes`, including the one in `reduce-bandwidths-mode`. Should we rather inherit the mode from the modes it enables, so that it's cleaner?
4. Should there be non-configurable default modes? Should `gopher`, `gemini` and internal page modes be redefinable from the `auto-rules.lisp`?
5. Should `enable` perform the work of `make-instance` automatically, or should it always be something more up to the one defining the method, like in `repeat-mode`?

# Checklist:
Everything in this checklist is required for each PR.  Please do not approve a PR that does not have all of these items.

- [X] I have pulled from master before submitting this PR
- [x] There are no merge conflicts.
- [X] I've added the new dependencies as:
  - [X] ASDF dependencies,
  - [X] Git submodules,
    ```sh
	cd /path/to/nyxt/checkout
    git submodule add https://gitlab.common-lisp.net/nyxt/py-configparser _build/py-configparser
    ```
  - [X] and Guix dependencies.
- [X] My code follows the style guidelines for Common Lisp code. See:
  - [Norvig & Pitman's Tutorial on Good Lisp Programming Style (PDF)](https://www.cs.umd.edu/~nau/cmsc421/norvig-lisp-style.pdf)
  - [Google Common Lisp Style Guide](https://google.github.io/styleguide/lispguide.xml)
- [X] I have performed a self-review of my own code.
- [x] My code has been reviewed by at least one peer.  (The peer review to approve a PR counts.  The reviewer must download and test the code.)
- [X] Documentation:
  - [X] All my code has docstrings and `:documentation`s written in the aforementioned style.  (It's OK to skip the docstring for really trivial parts.)
  - [x] I have updated the existing documentation to match my changes.
  - [X] I have commented my code in hard-to-understand areas.
  - [x] I have updated the `changelog.lisp` with my changes if it's anything user-facing (new features, important bug fix, compatibility breakage).
  - [x] I have added a `migration.lisp` entry for all compatibility-breaking changes.
- [X] Compilation and tests:
  - [X] My changes generate no new warnings.
    - More so: the "Cannot open-code test the type HISTORY-ENTRY" is gone!
  - [X] I have added tests that prove my fix is effective or that my feature works.  (If possible.)
  - [X] New and existing unit tests pass locally with my changes.

aartaka commented:
I've pushed a questionable change: `(en|dis)able-modes now uses keyword arguments, including the `:bypass-auto-rules-p` that's passed directly to `enable`/`disable`. It breaks quite some modes. Investigating. In meanwhile, does it look like a reasonable move to you?

aartaka commented:
OK, it doesn't hang anymore, but `repeat-mode` doesn't seem to work (on master too :D). Investigating...

aartaka commented:
OK, it looks more or less complete (except for `repeat-mode` being severely broken everywhere, even on `master`). And there are many useful changes, too. @Ambrevar, @aadcg, @jmercouris, care to review?

Ambrevar commented on source/mode.lisp:
```diff
@@ -86,6 +86,7 @@ The handlers take the mode as argument.")
 
 (export-always 'enable)
 (defgeneric enable (mode &key &allow-other-keys)
+  (:method-combination cascading)
```
Explain why we want this in the `:documentation`.

Ambrevar reviewed (commented):
Wow, impressive work!

I'll need to give this review a 2nd round.
Note to myself:
- How `pre-request-hook` is replaced.
- Individual commits.
- `define-auto-rule` logic.
- How `small-web` got so much shorter.

aartaka commented on source/renderer/gtk.lisp:
```diff
@@ -981,25 +981,29 @@ See `finalize-buffer'."
       (setf modifiers (funcall (modifier-translator *browser*)
                                (webkit:webkit-navigation-action-get-modifiers navigation-action))))
     (setf url (quri:uri (webkit:webkit-uri-request-uri request)))
+    ;; `toplevel-p'
```
I mean that it's equivalent to the `:toplevel-p` argument of `request-data`. I guess we have to add yet another let-binding :)

aartaka commented on source/renderer/gtk.lisp:
```diff
@@ -981,25 +981,29 @@ See `finalize-buffer'."
+    ;; `toplevel-p'
+    (when (quri:uri=
+           url (quri:uri (webkit:webkit-web-view-uri
+                          (gtk-object buffer))))
```
Yes, it is. It previously was part of `:toplevel-p` `request-data` slot calculation. And now this slot is only accessible later, so we can't rely on it.

aartaka replied:
OK, it's a `let`-bound variable now.

aartaka commented on source/mode.lisp:
```diff
@@ -385,8 +408,9 @@ If it's a single buffer, return it directly (not as a list)."
     (mapcar (lambda (buffer)
-              (mapcar #'disable (delete nil (mapcar (lambda (mode) (find mode (modes buffer) :key #'name))
-                                                    (uiop:ensure-list modes)))))
+              (mapcar (alex:rcurry #'disable :bypass-auto-rules-p bypass-auto-rules-p)
```
I dunno...

aartaka replied:
Both sound quite arbitrary, but bypass is as if we're trying to circumvent some gatekeeper, which auto-rules watching every request indeed are.

Ambrevar replied:
To repeat what I said elsewhere, while I understand your point and yes, there is a semantic nuance, I believe most users will be unaware of it and would prefer consistency in the wording.

aadcg commented on source/buffer.lisp:
```diff
@@ -157,6 +157,56 @@ Nothing to do for the simplest `buffer' type."
+    :documentation "The list of `mode-invocation's that were enabled on the last
+URL not covered by `auto-mode'.  This is useful when alternative between
+rule-less and ruled pages.  Example browse sequence:
+
```
Please re-read.

aadcg replied:
Ok, I finally guessed what you actually wanted to say!

You mean "alternating", not "alternative". I wouldn't say "rule-less and rules pages", but perhaps "between URLs for which auto-rules are enabled or disabled".

aadcg replied:
I believe that it makes no longer sense to refer to `auto-mode`, correct?

aartaka replied:
Yes.

aadcg commented on source/buffer.lisp:
```diff
@@ -157,6 +157,56 @@ Nothing to do for the simplest `buffer' type."
+
+In the above, when browsing from nyxt.atlas.engineer to en.wikipedia.org, the
+modes that were in place before the nyxt.atlas.engineer rule was applied are
+restored.")
```
Alas.

aadcg commented on source/user-classes.lisp:
```diff
@@ -100,3 +100,52 @@ In a later file, when you've defined `gtk-browser':
+   (around (:around))
+   (after (:after))
+   (primary ()))
+  "Cascade to all the parent-specialized methods after the primary one.
```
I like this. 

I believe the method combination name should be "cascade", not "cascading". 

This docstring is hard to follow for a number of reasons (although I know precisely what you want to say).

If you call your classes "a" and "b", then you have to somehow distinguish them from your prose (like I just did), otherwise it's a mess. 

Do you really need to write some pseudo-code to explain the idea? Perhaps you don't. Is this the shortest example that hides all distracting details? I don't think so.

It doesn't seems like a very good or efficient idea to write code and add comments on that code snippet in a docstring.

What is "DRY"? [Where does he comes from? What's he made of?](https://www.youtube.com/watch?v=s2HrNIluolo&t=80)

(You don't necessarily need to agree or change anything, but you get the point).

aartaka replied:
I've renamed the combination and removed the code examples. Not sure that makes the docstring any more understandable, though :)

aadcg replied:
> "Cascade to all the parent methods after the child one.
Allows for composed object constructors/destructors, for instance."

My suggestion:

"Cascade upwards in the hierarchy from the child to all parent methods."

Since you decided to take constructors (or destructors) as an example of its application, you could exhibit a concrete simple example with a constuctor.

aadcg commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,454 @@
+(defun make-mode-toggle-prompting-handler (enable-p buffer)
+  #'(lambda (mode)
+      (sera:and-let* ((applicable (not (bypass-auto-rules-p buffer)))
+                      (promptable (prompt-on-mode-toggle-p buffer))
```
Both `applicable` and `promptable` should be renamed to `_`. They're dummy variables and, at least in Scheme, it' s common to name those `_`.

aadcg commented on source/mode.lisp:
```diff
@@ -120,15 +131,29 @@ See also `disable'."))
 See also `enable'."))
 
+(defmethod disable :around ((mode mode) &key bypass-auto-rules-p &allow-other-keys)
+  (if bypass-auto-rules-p
```
`when`?

aartaka replied:
Why?

aartaka replied:
Note that there are two branches.

aadcg replied:
Ups, sorry. The nested `if`s don't bother me personally, but some don't fancy them.

aadcg commented on source/mode/repeat.lisp:
```diff
@@ -40,25 +46,30 @@ It takes a `repeat-mode' instance as argument.")))
-                       (prompt1 :prompt "Repeat every X seconds"
-                                :input "5"
-                                :sources 'prompter:raw-source))))))
+  (let* ((seconds (or seconds
```
Why `let*`? 

I'd refactor this with `sera:when-let`.

aadcg commented:
Obs:

- In commit `53aca2d23` there are changes that seem orthogonal to its description (namely changes to `small-web`);
- Commits like  `3c368fb4d`, `0acb67334` or `a0cbde45c` should always be squashed in my opinion (since it's part of your "internal" reasoning).

aadcg commented:
A very welcomed change! Now I'm going to test this. I didn't use `auto-mode` before, so I need some time to get used to this :)

aartaka commented:
> Obs:
> 
>  * In commit [53aca2d](https://github.com/atlas-engineer/nyxt/commit/53aca2d23b01bf8b3513f602b0fa67d1afca032c) there are changes that seem orthogonal to its description (namely changes to `small-web`);

No! The changes to `small-web` are exactly in line with the commit meaning. `small-web-mode` used to reproduce _a lot_ of auto-rules functions, and now it simply uses auto-rules!

>  * Commits like  [3c368fb](https://github.com/atlas-engineer/nyxt/commit/3c368fb4d2d9a5d0e477879e8fa585e313840aab), [0acb673](https://github.com/atlas-engineer/nyxt/commit/0acb673349ff02e0c2001218fa70e7c1a111d7b4) or [a0cbde4](https://github.com/atlas-engineer/nyxt/commit/a0cbde45cc37e3ba63a670715d0b36acd182c02b) should always be squashed in my opinion (since it's part of your "internal" reasoning).

I don't see them being much of a problem, but I guess you're right :)

aadcg commented on source/mode.lisp:
```diff
@@ -120,15 +131,29 @@ See also `disable'."))
 This method is meant to be specialized for every mode.
 It is not meant to be called directly, see `disable-modes' instead.
 
+All the parent modes' `didable' methods run after the exact mode one, cascading
```
Typo "didable".

aartaka replied:
Fixed in 895219ed3c60aa554c8d63502e7bd09abf86a120.

aadcg replied:
My suggestion:

```
"Run when disabling a mode.
The pre-defined `:after' method handles further cleanup.
This method is meant to be specialized for every mode.
It is not meant to be called directly, see `disable-modes' instead.

All of the parent MODEs' `disable' methods run sequentially, cascading upwards
in the hierarchy. See the `cascade' method combination for a description.

See the analogous `enable'."
```

aadcg commented:
I believe you still haven't deprecated the `rememberable-p` slot from `mode`, correct?

aadcg commented:
Please review `toggle-modes`, `https->http-loop-help` and the `config-fails-and-browser-restarts` test.

aadcg commented:
In the docstrings and symbol names you seem to mix and match "previous" and "last", as well as "page" and "url".

When referring to "last URL", it might perhaps be better to say "last visited URL" to avoid abusing language and confusing the users.

If they do refer to the same thing, can we settle on a unique definition?

aartaka commented:
> I believe you still haven't deprecated the `rememberable-p` slot from `mode`, correct?

No!

aartaka commented:
> Please review `toggle-modes`, `https->http-loop-help` and the `config-fails-and-browser-restarts` test.

All done, except `toggle-modes`. Seems like some complex logic there, need to get back to it later.

aartaka commented:
`toggle-modes` done :)

Ambrevar commented:
Sorry for the delay, I'll get back to you as soon as possible!

aartaka commented:
To do: remove `mode-invokation` for good now that mode symbols are reliable mode representations.

aartaka commented:
OK, 6485dca removes `mode-invocation`, but I'm not sure that's the way I want to see it. @aadcg, @Ambrevar, @jmercouris, how does it look to you?

aadcg commented:
I've left some minor comments as requested. Overall I trust your judgement, since you don't have the big picture that you have, @aartaka.

aartaka commented:
@Ambrevar, time for a second round of review for this :) The most important question is: should we stay with `mode-invocation` objects, or use mere lists with enhanced typing, as in 6485dcaf723c13ec46d61b48b869769607c475f0?

aartaka commented:
I've also rebased over master. It works, it fixes the linked issues, it is mergeable (unless there is some more feedback on a better design it could have).

Ambrevar commented:
OK, will review when I'm done with #2536.

aadcg commented:
@aartaka could you please rebase over master so that I can review this?

aartaka commented:
@aadcg, done! CI is failing on CCL, but the functionality works just fine :)

aadcg commented:
@aartaka what do you mean by "rebased syntax" in "Fix the rebased syntax for enable-modes." (commit fffee9fef)?

aadcg commented:
I've cut the number of commits significantly. 

Commit e1cf9ef4e should be reworded - what exactly is broken? The explanation should be added below the commit top message.

aadcg commented:
Let me learn how to use the new version of "auto-mode" and play with it for some hours and I'll comment on the functionality.

aartaka commented:
Sure!

aartaka commented:
> Commit [e1cf9ef](https://github.com/atlas-engineer/nyxt/commit/e1cf9ef4ed4959d6cacaab6fd5c7d31eda09c648) should be reworded - what exactly is broken? The explanation should be added below the commit top message.

I've reworded it in 59ed4e42b9f5a2192be82ba78d9978cfd08cb26a.

aadcg commented:
> mode/repeat: Make it most correct code possible (still BROKEN tho).
> 
> repeat-seconds is still broken. It was broken on master too, so it's not a
> fault—and not a part of the scope—of this patch, but it's still annoying.

What's `repeat-seconds`? Perhaps you meant the `repeat-every` command? 

It's still not clear what exactly is broken. Perhaps an example could be provided. 

Suggestion:

> 
> mode/repeat: Refactor repeat-every.
> 
> It's still unreliable for this and that reason.

aadcg commented:
Should this feature be baptized "Auto rules" or "Auto-rules"?

aadcg commented:
@aartaka see 
217d2f4,  as I've re-wrote the manual section.

aartaka commented:
> > mode/repeat: Make it most correct code possible (still BROKEN tho).
> > repeat-seconds is still broken. It was broken on master too, so it's not a
> > fault—and not a part of the scope—of this patch, but it's still annoying.
> 
> What's `repeat-seconds`? Perhaps you meant the `repeat-every` command?
> 
> It's still not clear what exactly is broken. Perhaps an example could be provided.

It simply doesn't work, silently failing somewhere へ(゜∇、°)へ That's why the description is opaque — it's simply broken, for whatever reason.

aartaka commented on source/manual.lisp:
```diff
@@ -224,7 +224,53 @@ all hosts being blocked, execute command " (:code "describe-variable") ", choose
+         (:li "Automatically, by setting "
+              (:nxref :slot 'prompt-on-mode-toggle-p :class-name 'modable-buffer)
+              " to non-nil (refer to the "
+              (:a :href "nyxt:manual#configuration" "configuration section")
```
You can use a mere `#configuration`, I believe.

aartaka commented on source/manual.lisp:
```diff
@@ -224,7 +224,53 @@ all hosts being blocked, execute command " (:code "describe-variable") ", choose
+              (:a :href "nyxt:manual#configuration" "configuration section")
+              " for help)."))
+        (:p "All rules are stored in the corresponding auto-rules file, "
+            (:code "(files:expand (make-instance 'auto-rules-file))")
```
Maybe make it an `:ncode` for testability? It will require re-flowing the paragraph, but will result in location being discoverable at runtime.

Maybe I should add an inline toggle for `:ncode`, though...

aadcg replied:
I tried that. What made be not use it is the fact that it introduces a new paragraph and a bunch of buttons to edit the code.

aartaka replied:
> I tried that. What made be not use it is the fact that it introduces a new paragraph 

That's changeable. Inline layout for `:ncode` will happen after we merge this, then.

>and a bunch of buttons to edit the code.

Buttons are nice and useful, especially for such a piece of code that's made to be evaluated and played with.

aadcg replied:
> > I certainly remember that at some point we had either the manual or the tutorial evaluate the location of init files, but it seems that it doesn't exist anymore.. Do you recall it?
> 
> There still is that part. See the `let` in `manual-sections`.

It all makes sense now. This was actually broken. See c20ba3b1a.

aadcg replied:
I'd stick to smth way simpler. See 842f8a2cd.

aartaka commented on source/manual.lisp:
```diff
@@ -224,7 +224,53 @@ all hosts being blocked, execute command " (:code "describe-variable") ", choose
+            ", which " (:u "is intended for read and write purposes")". You can
+find instructions at the top of mentioned file. The gist is that rules are
+mere Lisp lists which start with a condition that checks the URL. When
+conditions are met, modes are toggled. Besides user-defined conditions, the
```
"Besides user-defined conditions" sounds confusing. I guess it was refactored from the idea that one can define arbitrary conditions? If so, then it should rather be expanded, maybe after the `match-*` conditions are mentioned.

aadcg replied:
I think you have a point. Let me reword it.

aadcg commented:
@aartaka it seems I have spotted a bug. Set `prompt-on-mode-toggle-p` to `t` and try enabling a mode. I got the following:

```
The value
  NIL
is not of type
  PACKAGE
when binding *PACKAGE*
   [Condition of type TYPE-ERROR]

Restarts:
 0: [FORWARD-CONDITION] FILES::FORWARD-CONDITION
 1: [ABORT] abort thread (#<THREAD "NFiles worker" RUNNING {1028ACAA03}>)

Backtrace:
 0: ((:METHOD FILES:SERIALIZE (NYXT-PROFILE AUTO-RULES-FILE T)) #<unused argument> #<AUTO-RULES-FILE {1037B6AD63}> #<SB-SYS:FD-STREAM for "file /home/aadcg/.local/share/nyxt/auto-rules-tmpID7X8UGC.lisp" {..
      Locals:
        FILE = #<AUTO-RULES-FILE {1037B6AD63}>
        STREAM = #<SB-SYS:FD-STREAM for "file /home/aadcg/.local/share/nyxt/auto-rules-tmpID7X8UGC.lisp" {1028AE5C63}>
 1: ((:METHOD FILES:WRITE-FILE (FILES:PROFILE FILES:FILE)) #<NYXT-PROFILE {1007A0BFE3}> #<AUTO-RULES-FILE {1037B6AD63}> :DESTINATION #P"/home/aadcg/.local/share/nyxt/auto-rules-tmpID7X8UGC.lisp") [fast-me..
      Locals:
        FILES::DESTINATION = #P"/home/aadcg/.local/share/nyxt/auto-rules-tmpID7X8UGC.lisp"
        FILES:FILE = #<AUTO-RULES-FILE {1037B6AD63}>
        FILES:PROFILE = #<NYXT-PROFILE {1007A0BFE3}>
        STREAM = #<SB-SYS:FD-STREAM for "file /home/aadcg/.local/share/nyxt/auto-rules-tmpID7X8UGC.lisp" {1028AE5C63}>
 2: ((:METHOD FILES:WRITE-FILE (FILES:PROFILE FILES:GPG-FILE)) #<NYXT-PROFILE {1007A0BFE3}> #<AUTO-RULES-FILE {1037B6AD63}> :DESTINATION #P"/home/aadcg/.local/share/nyxt/auto-rules-tmpID7X8UGC.lisp") [fas..
      Locals:
        SB-PCL::.NEXT-METHOD-CALL. = #S(SB-PCL::FAST-METHOD-CALL :FUNCTION #<FUNCTION (SB-PCL::FAST-METHOD FILES:WRITE-FILE (FILES:PROFILE FILES:FILE))> :PV NIL :NEXT-METHOD-CALL NIL :ARG-INFO (2 . T))
        SB-PCL::.REST-ARG. = (:DESTINATION #P"/home/aadcg/.local/share/nyxt/auto-rules-tmpID7X8UGC.lisp")
        FILES:FILE = #<AUTO-RULES-FILE {1037B6AD63}>
        FILES:PROFILE = #<NYXT-PROFILE {1007A0BFE3}>
 3: ((LAMBDA (FILES::DESTINATION) :IN FILES:WRITE-FILE) #P"/home/aadcg/.local/share/nyxt/auto-rules-tmpID7X8UGC.lisp")
      Locals:
        #:G67 = #P"/home/aadcg/.local/share/nyxt/auto-rules-tmpID7X8UGC.lisp"
 4: (UIOP/STREAM:CALL-WITH-STAGING-PATHNAME #P"/home/aadcg/.local/share/nyxt/auto-rules.lisp" #<FUNCTION (LAMBDA (FILES::DESTINATION) :IN FILES:WRITE-FILE) {1028AE3BEB}>)
      Locals:
        FUN = #<FUNCTION (LAMBDA (FILES::DESTINATION) :IN FILES:WRITE-FILE) {1028AE3BEB}>
        PATHNAME = #P"/home/aadcg/.local/share/nyxt/auto-rules.lisp"
        PATHNAME#1 = #P"/home/aadcg/.local/share/nyxt/auto-rules.lisp"
        STAGING = #P"/home/aadcg/.local/share/nyxt/auto-rules-tmpID7X8UGC.lisp"
 5: ((:METHOD FILES:WRITE-FILE :AROUND (FILES:PROFILE FILES:FILE)) #<NYXT-PROFILE {1007A0BFE3}> #<AUTO-RULES-FILE {1037B6AD63}>) [fast-method]
      Locals:
        SB-PCL::.NEXT-METHOD-CALL. = #S(SB-PCL::FAST-METHOD-CALL ..)
        SB-PCL::.REST-ARG. = NIL
        FILES::ENTRY = #<FILES::CACHE-ENTRY {10063FB953}>
        FILES:FILE = #<AUTO-RULES-FILE {1037B6AD63}>
        FILES::GROUP = NIL
        FILES::PATH = #P"/home/aadcg/.local/share/nyxt/auto-rules.lisp"
        NFILES/PATHNAME:PERMISSIONS = NIL
        FILES:PROFILE = #<NYXT-PROFILE {1007A0BFE3}>
        FILES::USER = NIL
 6: (FILES::WORKER-WRITE #<AUTO-RULES-FILE {1037B6AD63}> #<FILES::CACHE-ENTRY {10063FB953}>)
      Locals:
        CACHE-ENTRY = #<FILES::CACHE-ENTRY {10063FB953}>
        FILE = #<AUTO-RULES-FILE {1037B6AD63}>
```

aadcg commented:
> It simply doesn't work, silently failing somewhere へ(゜∇、°)へ That's why the description is opaque — it's simply broken, for whatever reason.

Oh... then I'll open an issue about that.

aartaka commented:
Package thing fixed in 566dc43e333c86f5ba7792e8d6d7280de00fb120.

aadcg commented:
```
;; - Condition for rule activation. It is either (match-domain ...), (match-host
;;   ...), (match-regex ...), (match-port ...), some other condition you define,
;;   or a string.
```

What is meant here that it's possible to define a condition by a string?

aadcg commented:
The URL itself. All clear!

aadcg commented:
@aartaka can you explain me what's the rationale behind `define-auto-rule` and `*default-rules*`?

We let users define their predicates (when `match-*` aren't enough), which seems to be flexible enough.

Also is it a good idea to have rules both stored in a global dynamical variable and in the `auto-rules-file`?

aadcg commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,435 @@
+                     (cl-ppcre:regex-replace "www[0-9]?\\." (quri:uri-host url) ""))
+            `(match-domain ,(quri:uri-domain url))
+            `(match-host ,(quri:uri-host url)))
+        `(match-url ,(render-url url)))))
```
Could we refactor this function with `cond`?

aartaka replied:
We could, but that would be a really ugly `cond`...

aadcg commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,435 @@
+               (invocation-member
+                (alex:when-let* ((previous-url (previous-url buffer))
+                                 (matching-rules (matching-auto-rules previous-url buffer)))
+                  (alex:mappend #'included matching-rules))))))))
```
`cond`?

aadcg replied:
It wouldn't be shorted, but easier on humans to parse.

aadcg commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,435 @@
+- If it's getting disabled (not ENABLED-P) after being enabled by a rule on the previous page."
+  (let ((mode (normalize-mode mode)))
+    (flet ((invocation-member (list)
+             (member mode list :test #'mode=)))
```
Perhaps define the local function in a single line?

Ambrevar replied:
I would find it harder to read.  I find that the line break highlights where the function name and argument list lie.

aadcg commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,435 @@
+           (previous-url (previous-url buffer))
+           (previous-rules (when previous-url (matching-auto-rules previous-url buffer))))
+      (when (and rules previous-url (not previous-rules))
+        (save-last-active-modes buffer previous-url))
```
Perhaps move this statement to the `cond` below?

aartaka replied:
No. It's there for side effects.

aadcg replied:
I see.

aadcg commented on tests/executable/config.lisp:
```diff
@@ -78,9 +78,9 @@
-       ;; `buffer' does not have an `auto-mode-rules-file' slot.
-       (setf (auto-mode-rules-file buffer) (make-instance 'auto-mode-rules-file
-                                                          :base-path "/path/to/auto/rules")))
+       ;; `buffer' does not have an `auto-rules-file' slot.
```
This comment seems to be out-of-date.

aadcg commented on source/buffer.lisp:
```diff
@@ -162,6 +162,55 @@ Nothing to do for the simplest `buffer' type."
+We can't rely on the previous history entry because dead buffers and
+session-restored buffers may have a history with a previous URL matching the
+same rule while obviously the rule has never been applied for the new-born
+buffer.")
```
This valuable comment seems to be better suited as a comment in the codebase and not as part of a docstring. Can we move it?

aartaka replied:
This is an unexported slot, and its presence should be justfied. Docstring is enough of a documentation, as those that get to the documentation of the unexported slot are likely the ones that need it there.

aadcg replied:
I didn't notice that it's unexported. All good.

aadcg commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,430 @@
+  (let ((rules (files:content file)))
+    (let ((*standard-output* stream)
+          (*package* (find-package :nyxt)))
+      (write-string ";;; List of auto-rules (meant to be readable and writable)
```
I refactored these instructions so that they don't repeat the information given in the manual.

I think that the information provided here should be succinct; one or two example should be given; and it should send the user to the manual for more info. 

Do you agree?

aartaka replied:
NOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO!

aartaka replied:
This file is probably the most thought-out Nyxt data file, exactly because it has _all_ of the instructions for how to use it—right at the top.

Yes, this information is duplicate of the info in the manual, but this file is to be opened in a text editor and edited there, possibly without the manual ever being in the loop. This commentary is the best documentation such a file could have.

aadcg replied:
I hear you. I'll bring it back.

aartaka commented:
> @aartaka can you explain me what's the rationale behind `define-auto-rule` and `*default-rules*`?

The huge problem of auto-mode before this patch was that we somehow had to force our own rules (without which Nyxt becomes unusable), while allowing the user to edit and add their own rules too. Pre-#2490 auto-mode only had editing and addition, but not mandatory default rules (the surrogate for those was basically toggling the page-related of scheme-related mode on in `request-resource-hook` and hoping it'd get disabled sometime after, which more often than not it wasn't).

Having everything put into one file, let alone somehow updating this file when we update the list of default rules, is hard if not impossible. More so: having it put into one file with user-editable rules implies that default rules are to be edited and deleted at will. Which is dangerous too.

That's why the solution was to introduce a piece of state called `*default-auto-rules*`, so that the rules that are required for Nyxt functioning are recognized as auto-rules, while being relatively un-discoverable for the user that doesn't need to know about those. I mean, they are still editable through REPL or SLIME/SLY connection, but that's significantly harder than just deleting a bunch of lines in auto-rules.lisp.

> We let users define their predicates (when `match-*` aren't enough), which seems to be flexible enough.

It's flexible, yes, but the problem is not in users defining rules. The problem is in _us_ and extension writers defining rules and needing them to be enforced.

> Also is it a good idea to have rules both stored in a global dynamical variable and in the `auto-rules-file`?

Yes, it it. It's two different purposes and thus implementations.

Ambrevar commented:
There is a load error reported by the CI:

```log
; wrote /home/runner/.cache/common-lisp/sbcl-2.2.4-linux-x64/home/runner/work/nyxt/nyxt/source/mode/repl-tmpWDB4LKR2.fasl
; compilation finished in 0:00:00.139
; 
; compilation unit aborted
;   caught 1 fatal ERROR condition
;   caught 1 WARNING condition
;   caught 66 STYLE-WARNING conditions
;   printed 3870 notes
caught error #<TYPE-ERROR expected-type: NYXT:BUFFER datum: NIL>
The value
  NIL
is not of type
  NYXT:BUFFER
when binding NYXT:BUFFER
```

@aartaka Any idea?

aartaka commented:
> There is a load error reported by the CI:
> 
> ```
> ; wrote /home/runner/.cache/common-lisp/sbcl-2.2.4-linux-x64/home/runner/work/nyxt/nyxt/source/mode/repl-tmpWDB4LKR2.fasl
> ; compilation finished in 0:00:00.139
> ; 
> ; compilation unit aborted
> ;   caught 1 fatal ERROR condition
> ;   caught 1 WARNING condition
> ;   caught 66 STYLE-WARNING conditions
> ;   printed 3870 notes
> caught error #<TYPE-ERROR expected-type: NYXT:BUFFER datum: NIL>
> The value
>   NIL
> is not of type
>   NYXT:BUFFER
> when binding NYXT:BUFFER
> ```
> 
> @aartaka Any idea?

No, no idea ｜−・;）

Ambrevar commented:
Maybe bisect to find out which commit is the culprit?

aadcg commented:
> > Also is it a good idea to have rules both stored in a global dynamical variable and in the `auto-rules-file`?
> 
> Yes, it it. It's two different purposes and thus implementations.

Thank you for the detailed explanation but I'm afraid I'm still a bit confused. I guess my question is as follows: the rules defined via `define-auto-rule` land on `*default-rules*`, while those defined via the auto rules files don't. I understand we need both mechanisms to add rules, but shouldn't the global variable track the global state? 

Also, perhaps `*default-rules*` should be renamed to `*default-auto-rules*`. (I have assumed it was called like that in a post above, sorry).

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,461 @@
+         (reapply-last-active-modes buffer))
+        ((and rules (not (eq rules previous-rules)))
+         (enable-matching-modes url buffer)))
+      (setf (previous-url buffer) url))))
```
Rules are computed here from `matching-auto-rules`, but then they are not forwarded down the call stack.  `enable-matching-modes` recalls `matching-auto-rules` in fact.  

I suppose this could be refactored a bit.

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,461 @@
+                (and (not enable-p) (invocation-member (alex:mappend #'excluded matching-rules)))))
+          ;; Mode is covered by auto-rules only if it is both in
+          ;; last-active-modes and gets enabled.  If it gets disabled, user
+          ;; should be prompted, because they may want to persist it.
```
To persist what?  How?

aartaka replied:
The mode being disabled. I've clarified the comment.

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,461 @@
+ENABLE-P is whether mode is being enabled (non-nil) or disabled (nil).
+
+Mode is covered if:
+- `rememberable-p' is to nil.
```
Typo?  "is nil" maybe?

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,461 @@
+
+Mode is covered if:
+- `rememberable-p' is to nil.
+- There is a rule it matches and:
```
"a matching rule"

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,461 @@
+(define-command-global save-non-default-modes-for-future-visits ()
+  "Save the modes present in `default-modes' and not present in current modes as
+:excluded, and modes that are present in mode list but not in `default-modes' as
+:included, to one of auto-rules. Apply the resulting rule for all the future
```
"to one of the auto-rules."

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,461 @@
+
+This command does not save non-rememberable modes. If you want auto-rules to
+remember a particular mode, configure it to be `rememberable-p' in your
+initfile.
```
"configuration file".

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,461 @@
+remember a particular mode, configure it to be `rememberable-p' in your
+initfile.
+
+For the storage format see the comment in the head of your `auto-rules-file'."
```
"header"

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,461 @@
+Uses `url-infer-match', see its documentation for matching rules.
+
+This command does not save non-rememberable modes. If you want auto-rules to
+save a particular mode, configure it to be `rememberable-p' in your initfile.
```
"configuration file"

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,461 @@
+
+For the storage format see the comment in the head of your `auto-rules-file'."
+  ;; TODO: Should it prompt for modes to save?
+  ;; One may want to adjust the modes before persisting them as :exact-p rule.
```
Maybe we need a graphical auto-mode settings page?  Should be easier one #2408 is merged.  In any case, this can wait.

aartaka replied:
Yes, graphical settings won't hurt, but they are beyond both this commentary and this PR.

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,461 @@
+                        (make-instance 'global-history-source
+                                       :return-actions '())))))
+    (when (typep url 'nyxt::history-entry)
+      (setf url (url url)))
```
I guess you can remove the condition here and just call `(setf url (url url))`, because the `url` function is idempotent.

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,461 @@
+
+;; - match-regex works for any address that matches a given regex. You can add
+;;   these manually, but remember: with great regex comes great responsibility!
+;;   Example: (match-regex \"https://github\\.com/.*/.*\") will activate only in
```
"activate" -> "match"

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,461 @@
+;;   URL as an argument. This means you can write arbitrary Lisp code to
+;;   activate auto-rules.
+;;   Note: The URL is passed as quri:uri object.
+;;   Example: (lambda (url) (string= \"/my/path\" (quri:uri-path url)))
```
Wait, is it a form or a lambda?

aartaka replied:
It's a lambda, I've unlocked it. One can now write arbitrary Lisp code as rule's test.

Ambrevar replied:
Then update the comment a few lines above, it says "form".

aartaka replied:
Oh wait, it's written just right: it can be any Lisp form _given that it evaluates to something that can be funcalled with the single URL argument_.

Ambrevar replied:
Oh, right.

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,461 @@
+  (let ((rules (files:content file)))
+    (let ((*standard-output* stream)
+          (*package* (find-package :nyxt)))
+      (write-string ";;; List of auto-rules (meant to be readable and writable)
```
Maybe "human-readable and human-writable"

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,461 @@
+    (let ((*standard-output* stream)
+          (*package* (find-package :nyxt)))
+      (write-string ";;; List of auto-rules (meant to be readable and writable)
+
```
Nit: I would start all empty lines with `;;` since it's all part of the same comment.

Ambrevar commented on source/buffer.lisp:
```diff
@@ -188,9 +240,12 @@ of BUFFER."
+  (enable-modes :modes (append (reverse (default-modes buffer))
+                               (uiop:ensure-list extra-modes))
+                :buffers buffer
+                ;; XXX: We don't want to trigger auto-rules when enabling
```
We don't use `XXX` much.  I propose to stick to `TODO`.

Ambrevar commented on source/buffer.lisp:
```diff
@@ -188,9 +240,12 @@ of BUFFER."
+                               (uiop:ensure-list extra-modes))
+                :buffers buffer
+                ;; XXX: We don't want to trigger auto-rules when enabling
+                ;; extra-modes, don't we?
```
"do we".

Ambrevar replied:
My answer: I don't see why interactive mode activation should trigger auto-rules, so I believe this is right.

Ambrevar commented on source/changelog.lisp:
```diff
@@ -471,7 +471,18 @@ SLY install.")
+specific one.")
+    (:li "There are default rules for Gopher, Gemini, and Nyxt-internal-pages.")
+    (:li "The rules file is now moved to auto-rules.lisp (instead of the old
+auto-mode-rules.lisp)")))
```
Actually it was in `mode/auto.lisp`.

aartaka replied:
You're talking about the file with auto-mode/auto-rules sources.
I'm talking about the data file with the rules themselves.

Ambrevar replied:
Oh right.

Ambrevar commented on source/manual.lisp:
```diff
@@ -15,9 +15,10 @@ of Nyxt."))
   (spinneret:with-html-string
+    (let ((auto-config-file (namestring (files:expand *auto-config-file*)))
+            (config-file (namestring (files:expand *config-file*)))
+            (rules-file (namestring (files:expand (make-instance 'auto-rules-file)))))
```
Indentation.

Ambrevar commented on source/buffer.lisp:
```diff
@@ -162,6 +162,58 @@ Nothing to do for the simplest `buffer' type."
+   (bypass-auto-rules-p
+    nil
+    :type boolean
+    :documentation "Whether to bypass auto-rules.")
```
Could you had an example?  It's not very obvious how this is meant to be used.

aartaka replied:
This is meant to be used when one wants to enable/disable modes, but doesn't want to trigger auto-rules, especially the interactive version of those.

I'll duplicate this to the docstring.

Ambrevar commented on source/mode.lisp:
```diff
@@ -86,23 +86,34 @@ The handlers take the mode as argument.")
+              (setf (bypass-auto-rules-p (buffer mode)) t)
+              (unwind-protect
+                   (call-next-method)
+                (setf (bypass-auto-rules-p (buffer mode)) old-bypass)))
```
This unwind-protect business is obviously not very pretty.

But why do we need to check `bypass-auto-rules-p` here?

It's not very clear to me when and why we need to bypass auto-rules.

Ambrevar commented on source/mode.lisp:
```diff
@@ -120,15 +131,29 @@ See also `disable'."))
+        (setf (bypass-auto-rules-p (buffer mode)) t)
+        (unwind-protect
+             (call-next-method)
+          (setf (bypass-auto-rules-p (buffer mode)) old-bypass)))
```
Same here as with `enable`.

Ambrevar commented on source/mode.lisp:
```diff
@@ -322,17 +347,22 @@ For production code, see `find-submode' instead."
+  (declare (ignorable bypass-auto-rules-p))
   ;; We allow NIL values for MODES and BUFFERS in case they are forms, in which
   ;; case it's handy that this function does not error, it simply does nothing.
+  ;; REVIEW: But we wrap commands into `with-protect' for this, don't we?
```
We do, but it would rewind the stack of the local thread and thus not run what's after it.
This was the source of undetected bugs for a while.

In any case, this part of the API should be improved.

Ambrevar replied:
Is this comment still relevant?

Ambrevar commented:
@aartaka You said it fixes #2064, but does it restore modes that were enabled manually while the rule was _not_ persisted?  This was the core issues of #2064.

Ambrevar commented on source/renderer-script.lisp:
```diff
@@ -216,6 +218,14 @@ See `find-internal-page-buffer'."))
                   (values contents type status))))))))
 
+(defmethod (setf page-mode) (new-value (page internal-page))
+  ;; FIXME: Better rule?
```
You mean that `str:starts-with-p` is too naive?

aartaka replied:
Yes. If we can avoid stringly-typed-programming, we should.

Ambrevar replied:
OK, please clarify the comment (also move it right above the str:starts-with-p call)

Ambrevar commented on source/urls.lisp:
```diff
@@ -387,7 +387,6 @@ guarantee of the same result."
               (parse-nyxt-url url)
             (when (and internal-page-name)
               (alex:when-let ((internal-page (gethash internal-page-name *nyxt-url-commands*)))
-                (enable-modes (page-mode internal-page) buffer)
```
Why should this go?  What's the better place to enable modes now?

aartaka replied:
The `(setf page-mode)` method adds the auto-rule for the mode now.

Ambrevar replied:
I see this:

```lisp
(defmethod (setf page-mode) (new-value (page internal-page))
  ;; FIXME: Better rule?
  (when (and new-value (rememberable-p new-value))
    (define-auto-rule `(lambda (url)
                         (str:starts-with-p ,(nyxt-url (name page))
                                            (render-url url)))
      :included (list new-value))))
```

So it defines the auto-rule, but it does not enable the modes, does it?

aartaka replied:
I does not. The code in `on-signal-decide-policy` does, 
- based on the rules found by `matching-auto-rules`, 
- which relies on `*default-auto-rules*`,
- where `define-auto-rule` pushes.

Ambrevar commented:
Alrighty, I gave this a second review and I really like it!

Left to answer (for me):
- I find the `bypass-...` logic suspicious.
- `*default-modes*` seems unused, or am I mistaken?

I'll do more testing tomorrow.  In the mean time, please address my comments, then we should be good to merge!

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,461 @@
+  (rememberable-p (make-instance mode)))
+
+(deftype mode-invocation ()
+  `(cons mode-symbol *))
```
While this `mode-invocation` type is nice and short, how does it handle mode arguments?
Previously, the structure supported arguments which would be passed to `make-instance` when the mode was first instantiated.

aartaka replied:
The arguments are still there, as a `rest` of the cons. This is not nice and it'd be nice to change it to something stricter. But it essentially is the same as the `arguments` slot typed as `list`, so there's not much hurry :)

Ambrevar replied:
OK, please document this, as it was unclear to me when I read this.

aartaka commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,461 @@
+    (princ (sera:ellipsize (format nil "~a" (test rule)) 40) stream)))
+
+(defvar *default-rules* '()
+  "The list of rules to always use, even when there is no auto-rules file.")
```
Not exported, because it's an internal detail of `define-auto-rule` that is intended strictly for internal use. Anything user-facing goes through `define-auto-rule` and Nyxt UI; anything developer-facing goes through `matching-auto-rules`.

aartaka commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,461 @@
+    (if (apply-all-matching-auto-rules-p buffer)
+        matching-rules
+        (flet ((priority (test1 test2)
+                 (let ((priority-list '(match-regex match-url match-host match-domain match-scheme)))
```
I't mostly just an intuition:
- Regex is the last option one has for granular matching, so we put it as highest priority.
- `match-url` matches just one page, so it follows regex.
- And then, host, domain, and sheme all go in the decreasing order of specificity and priority. 
It's made this way so that more specific rules can override/circumvent the less specific rules.

Ambrevar replied:
OK, you can add this as a comment maybe.

But what if we want to add a new `match-*` some day?  And what if the user wants to extend this themselves?

aartaka replied:
I've added a docstring to `priority` in 70d987027ceed0849f86e894a01559d8035aff9e.

If we add a new `match-*` (and we already did), we either add it to the list or do nothing. This list is mainly to emphasize the default predicates as more reliable than user-defined ones, and to make sulre user-defined predicates are only used when  absolutely necessary. And to have a way to resolve conflicts between different matching rules too—that was the initial intention :)

aartaka commented:
> * I find the `bypass-...` logic suspicious.

I mean, yeah, it's suspicious, but that seemed the only way to supress a newly global might of auto-rules that intercept any mode toggle and any request now ヽ(゜Q。)ノ？

> * `*default-modes*` seems unused, or am I mistaken?

It's used in `matching-auto-rules`.

> * Does it really fix [Modes are not restored when restoring session #2064](https://github.com/atlas-engineer/nyxt/issues/2064), i.e. are modes saved in the GHT somehow?

No, it only fixes it for the pages with auto-rules. I'll remove this issue from the list of linked issues.

> * Can `mode-invocation` still pass arguments to `make-instance` when the mode is first instantiated?

Yes, it can.

aartaka commented:
> > > Also is it a good idea to have rules both stored in a global dynamical variable and in the `auto-rules-file`?
> > 
> > 
> > Yes, it it. It's two different purposes and thus implementations.
> 
> Thank you for the detailed explanation but I'm afraid I'm still a bit confused. I guess my question is as follows: the rules defined via `define-auto-rule` land on `*default-rules*`, while those defined via the auto rules files don't. I understand we need both mechanisms to add rules, but shouldn't the global variable track the global state?

It doesn't have to. In our case, the global state is split between:
- Nyxt-internal, hidden, default, non-overridable `*default-rules*`, and
- public, editable, file-backed `auto-rules-file`.
Those converge in `matching-auto-rules` to form the final version of rule list. Before that, the user-specific and Nyxt-default rules _shall not_ be conflated in any way, as that would either:
- allow the user to edit default rules breaking some parts of Nyxt, or
- spill some default rules into the auto-rules file, making them duplicated and buggy.

> Also, perhaps `*default-rules*` should be renamed to `*default-auto-rules*`. (I have assumed it was called like that in a post above, sorry).

Could be. But, given that it's unexported, it's not that critical.

I'll rename them.

Ambrevar commented:
> > * `*default-modes*` seems unused, or am I mistaken?
> 
> It's used in `matching-auto-rules`.

I see that, but it's always empty (and it's unexported).  So what's the point?

Ambrevar commented:
> > *
> 
> I mean, yeah, it's suspicious, but that seemed the only way to supress a newly global might of auto-rules that intercept any mode toggle and any request now ヽ(゜Q。)ノ？

Can you give us an overlook of the new logic here?  Where does the interception happen for instance?

aartaka commented:
> > > * `*default-modes*` seems unused, or am I mistaken?
> > 
> > 
> > It's used in `matching-auto-rules`.
> 
> I see that, but it's always empty (and it's unexported). So what's the point?

It's not—see `define-auto-rule` that pushes directly to `*default-auto-rules*`.

aartaka commented:
TODO: split `(dis|en)able-modes` command into two parts:
- Low-level mode toggling function not invoking auto-rules.
- High-level command, invoking auto-rules (and auto-rule-related prompting) when necessary.

aartaka commented:
> There is a load error reported by the CI:
> 
> ```
> ; wrote /home/runner/.cache/common-lisp/sbcl-2.2.4-linux-x64/home/runner/work/nyxt/nyxt/source/mode/repl-tmpWDB4LKR2.fasl
> ; compilation finished in 0:00:00.139
> ; 
> ; compilation unit aborted
> ;   caught 1 fatal ERROR condition
> ;   caught 1 WARNING condition
> ;   caught 66 STYLE-WARNING conditions
> ;   printed 3870 notes
> caught error #<TYPE-ERROR expected-type: NYXT:BUFFER datum: NIL>
> The value
>   NIL
> is not of type
>   NYXT:BUFFER
> when binding NYXT:BUFFER
> ```
> 
> @aartaka Any idea?

I found it! `nyxt/repl-mode:repl` has a `(find-submode :repl-mode repl-buffer)` that gets invoked at compile-time and causes errors with `(current-buffer)` being unbound. That's highly weird. However, I found ways to fix it, and will likely push a fix soon.

aartaka commented:
Okay, I've split `enable-modes` into:
- low-level `enable-modes*` helper that only toggles modes,
- and a high-level `enable-modes` that prompts the user, chooses several modes and buffers, and invokes prompting auto-rules handler whenever necessary.

Same for `disable-modes`. I've also added the new `prompt-on-mode-toggle` function (responsible for prompting the user about the modes being enabled/disabled and whether to save them all) to `toggle-modes`.

It all should look and work much more elegant now, but I'll test it some more tomorrow just in case.

aadcg commented:
Taking a final look now. Might take some time :)

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,463 @@
+(defmethod rememberable-p ((mode symbol))
+  (check-type mode mode-symbol)
+  ;; FIXME: Can this finalization shoot us in the foot?
+  (closer-mop:ensure-finalized (find-class mode) nil)
```
Why would you ensure finalization anyways?

aartaka replied:
Because we cannot call `mopu:class-slots` on a non-finalized class!

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,463 @@
+- MATCH-REGEX is the last option one has for granular matching, so we put it as
+  highest priority.
+- MATCH-URL matches just one page, so it follows regex in granularity.
+- And then, host, domain, and sheme all go in the decreasing order of
```
Typo "scheme"

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,463 @@
+
+(define-command-global save-non-default-modes-for-future-visits ()
+  "Save the modes present in `default-modes' and not present in current modes as
+:excluded, and modes that are present in mode list but not in `default-modes' as
```
Maybe mention that `default-modes' is a slot that belongs to `modable-buffer`.

aartaka replied:
No, because both our `resolve-backtick-quote-links` and Emacs' xref are pretty obvious about this reference being a method/slot. It's clear enough.

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,463 @@
+Uses `url-infer-match', see its documentation for matching rules.
+
+This command does not save non-rememberable modes. If you want auto-rules to
+save a particular mode, configure it to be `rememberable-p' in your
```
Maybe give an example?

Ambrevar commented on source/buffer.lisp:
```diff
@@ -162,6 +162,54 @@ Nothing to do for the simplest `buffer' type."
+    nil
+    :type boolean
+    :documentation "Whether matching auto-rules are applied to the URL.
+By default, the most specific rule is applied.")
```
Replace "by default" with "If nil".

Ambrevar commented on source/buffer.lisp:
```diff
@@ -162,6 +162,54 @@ Nothing to do for the simplest `buffer' type."
+    nil
+    :export nil
+    :type (or quri:uri null)
+    :documentation "The last URL that the active modes were saved for.
```
"The last URL for which the active modes were saved."

Ambrevar commented on source/auto-rules.lisp:
```diff
@@ -0,0 +1,463 @@
+       `(match-host ,(quri:uri-host url)))
+      (t `(match-url ,(render-url url))))))
+
+(defun prompt-on-mode-toggle (modes buffers enable-p)
```
Maybe make `enable-p` a keyword, so that call sites are clearer.
Example: `(prompt-on-mode-toggle MODES BUFFERS :enabled-p t)`.

Ambrevar commented:
Looking really good!
I left a few nit-picky comments.

My last concern: can we find better names for `enable-modes` and `enable-modes*`?

aartaka commented:
> My last concern: can we find better names for `enable-modes` and `enable-modes*`?

Indeed, this one is a placeholder-ish...

aadcg commented:
I've squashed some commits while maintaining everything as it was. It wouldn't hurt to check against your local branch.

Personally I think we should squash for instance a7176243e onto cf5c5e95a, but it's not a simple squash for some reason and my git skills aren't where they should be.

aadcg commented:
[I'm also wondering about this](https://english.stackexchange.com/questions/598057/is-all-matching-pleonasm).

aadcg commented:
> This also adds bypass-auto-rules-p argument to bypass auto-rules for one mode enabling session.

@aartaka could you explain what you mean by "for one mode enabling session'?

aartaka commented:
> > This also adds bypass-auto-rules-p argument to bypass auto-rules for one mode enabling session.
> 
> @aartaka could you explain what you mean by "for one mode enabling session"?

Wait, this part shouldn't be there, as `bypas-auto-rules-p` is removed now!

But what it meant was that it disables auto-mode for exactly one mode toggling. Like, for one `enable-mode`, or `disable-mode`, or `toggle-mode`.

aadcg commented:
I'm looking at the message from commit 33f977aab2bdbcdc2a423bc02f54fdf3a8ecd799. If you think it should be changed, could you please write it down here? I'll reword it and keep you as the author of the commit.

aadcg commented:
I have now rebased it and improved the commit log.

@aartaka can you run this branch and make a final check of the functionality?

aartaka commented:
> I have now rebased it and improved the commit log.
> 
> @aartaka can you run this branch and make a final check of the functionality?

Seems to work jsut fine for me! Some modes don't get toggled on link clicks, but link clicks have been problematic all along (even with the old auto-mode), so I doubt we can do better there :)

aadcg commented:
```log
; file: /home/aadcg/common-lisp/nyxt/source/mode/remembrance.lisp
; in: DEFUN NYXT/REMEMBRANCE-MODE::URL->CACHE
;     (NYXT:ENABLE-MODES 'NYXT/REMEMBRANCE-MODE:REMEMBRANCE-MODE
;                        NYXT:BACKGROUND-BUFFER)
; 
; caught STYLE-WARNING:
;   NYXT/REMEMBRANCE-MODE:REMEMBRANCE-MODE is not a known argument keyword.


; file: /home/aadcg/common-lisp/nyxt/source/mode/repeat.lisp
; in: NYXT:DEFINE-COMMAND-GLOBAL NYXT/REPEAT-MODE:REPEAT-TIMES
;     (NYXT:ENABLE-MODES :MODES 'NYXT/REPEAT-MODE:REPEAT-MODE :BUFFERS
;                        (NYXT:CURRENT-BUFFER) :REPEAT-COUNT
;                        NYXT/REPEAT-MODE::TIMES :REPEAT-ACTION
;                        #'(LAMBDA (NYXT:MODE)
;                            (DECLARE (IGNORE NYXT:MODE))
;                            (NYXT::RUN FUNCTION)))
; 
; caught STYLE-WARNING:
;   :REPEAT-COUNT and :REPEAT-ACTION are not a known argument keywords.

; in: NYXT:DEFINE-COMMAND-GLOBAL NYXT/REPEAT-MODE:REPEAT-EVERY
;     (NYXT:ENABLE-MODES :MODE 'NYXT/REPEAT-MODE:REPEAT-MODE :BUFFERS
;                        (NYXT:CURRENT-BUFFER) :REPEAT-INTERVAL
;                        NYXT/REPEAT-MODE::SECONDS :REPEAT-ACTION FUNCTION)
; 
; caught STYLE-WARNING:
;   :MODE, :REPEAT-INTERVAL and :REPEAT-ACTION are not a known argument keywords.
; 
; compilation unit finished
;   caught 3 STYLE-WARNING conditions
;   printed 4 notes
```

Could you please take a look at these compiler warnings?

aadcg commented:
> Some modes don't get toggled on link clicks, but link clicks have been problematic all along (even with the old auto-mode)

No worries! Do we have an issue on that?

aartaka commented:
> > Some modes don't get toggled on link clicks, but link clicks have been problematic all along (even with the old auto-mode)
> 
> No worries! Do we have an issue on that?

I remember it being reported somewhere, but I can no longer find it :( #2634 is a new one then!

aartaka commented:
> ```
> ; file: /home/aadcg/common-lisp/nyxt/source/mode/remembrance.lisp
> ; in: DEFUN NYXT/REMEMBRANCE-MODE::URL->CACHE
> ;     (NYXT:ENABLE-MODES 'NYXT/REMEMBRANCE-MODE:REMEMBRANCE-MODE
> ;                        NYXT:BACKGROUND-BUFFER)
> ; 
> ; caught STYLE-WARNING:
> ;   NYXT/REMEMBRANCE-MODE:REMEMBRANCE-MODE is not a known argument keyword.
> 
> 
> ; file: /home/aadcg/common-lisp/nyxt/source/mode/repeat.lisp
> ; in: NYXT:DEFINE-COMMAND-GLOBAL NYXT/REPEAT-MODE:REPEAT-TIMES
> ;     (NYXT:ENABLE-MODES :MODES 'NYXT/REPEAT-MODE:REPEAT-MODE :BUFFERS
> ;                        (NYXT:CURRENT-BUFFER) :REPEAT-COUNT
> ;                        NYXT/REPEAT-MODE::TIMES :REPEAT-ACTION
> ;                        #'(LAMBDA (NYXT:MODE)
> ;                            (DECLARE (IGNORE NYXT:MODE))
> ;                            (NYXT::RUN FUNCTION)))
> ; 
> ; caught STYLE-WARNING:
> ;   :REPEAT-COUNT and :REPEAT-ACTION are not a known argument keywords.
> 
> ; in: NYXT:DEFINE-COMMAND-GLOBAL NYXT/REPEAT-MODE:REPEAT-EVERY
> ;     (NYXT:ENABLE-MODES :MODE 'NYXT/REPEAT-MODE:REPEAT-MODE :BUFFERS
> ;                        (NYXT:CURRENT-BUFFER) :REPEAT-INTERVAL
> ;                        NYXT/REPEAT-MODE::SECONDS :REPEAT-ACTION FUNCTION)
> ; 
> ; caught STYLE-WARNING:
> ;   :MODE, :REPEAT-INTERVAL and :REPEAT-ACTION are not a known argument keywords.
> ; 
> ; compilation unit finished
> ;   caught 3 STYLE-WARNING conditions
> ;   printed 4 notes
> ```
> 
> 
>     
>   
> 
> Could you please take a look at these compiler warnings?

Fixed!

aadcg commented:
Beautiful, thanks!

aartaka commented:
Wohoooooooooo, finally!
